src/natcap/invest/validation.py

```diff
@@ -905,111 +913,73 @@ def validate(args, spec, spatial_overlap_opts=None):
-    # present in args and True is sufficient.
-    insufficient_keys = missing_keys.union(keys_with_no_value)
-    sufficient_inputs = {}
-    for key, parameter_spec in spec.items():
-        # If the key isn't present, no need to validate.
-        # If it's required and isn't present, we wouldn't have gotten to this
-        # point in the function.
-        if key not in args:
-            sufficient_inputs[key] = False
-            insufficient_keys.add(key)
-            continue
+        validation_warnings.append(
+            (sorted(keys_with_no_value), MESSAGES['MISSING_VALUE']))
 
-        # If the value is empty and it isn't required, then we don't need to
-        # validate it.
-        if args[key] in ('', None):
-            sufficient_inputs[key] = False
-            insufficient_keys.add(key)
-            continue
-
-        # Boolean values are special in that their T/F state is equivalent
-        # to their satisfaction.  If a checkbox is checked, it is
-        # considered satisfied.
-        if spec[key]['type'] == 'boolean':
-            sufficient_inputs[key] = args[key]
-
-        # Any other input type must be sufficient because it is in args and
-        # has a value.
-        else:
-            sufficient_inputs[key] = True
-
-    # step 3: evaluate required status of conditionally required keys
-    # keep track of keys that are explicity not required due to
-    # their condition being false
-    excluded_keys = set()
-    for key in conditionally_required_keys:
-        # An input is conditionally required when the expression given
-        # evaluates to True.
-        is_conditionally_required = _evaluate_expression(
-            expression=spec[key]['required'],
-            variable_map=sufficient_inputs)
-        if is_conditionally_required:
-            if key not in args:
-                validation_warnings.append(([key], MESSAGES['MISSING_KEY']))
-            else:
-                if args[key] in ('', None):
-                    validation_warnings.append(([key], MESSAGES['MISSING_VALUE']))
-        else:
-            excluded_keys.add(key)
-
-    # step 4: validate keys, but not conditionally excluded ones.
-    # Making a distinction between keys which are optional (required=False),
-    # and keys which are conditionally not required
-    # (required="condition that evaluates to False")
-    # We want to do validation on optional keys, like `n_workers`,
-    # but not on conditionally excluded keys, like fields that are greyed out
-    # because a checkbox is unchecked.
+    # Phase 2: Check whether any input with a value validates with its
+    # type-specific check function.
     invalid_keys = set()
-    sufficient_keys = set(args.keys()).difference(insufficient_keys)
-    for key in sufficient_keys.difference(excluded_keys):
+    insufficient_keys = (missing_keys | keys_with_no_value)
+    for key in set(args.keys()) - insufficient_keys:
```

Comment:
If I'm understanding, there is a subtle change here: before, we removed `excluded_keys` from the set to validate. Now, any keys that have a value will be validated, even if they're not required. I'm not sure what the correct answer is, but it seems worth noting the change.

Reply:
Yes, this exclusion was definitely something I needed to address, and the fix now includes a better variable name for `excluded_keys` to clarify how they were being used.

Reply:
@phargogh might there still be a difference if a non-required key has a real value? Suppose you have a `valuation_table` that's required only if `do_valuation` is True. Then if we provide args where `do_valuation` is False and `valuation_table` is a valid path, should we still validate `valuation_table`?

Reply:
Great question, and thank you for thinking critically about this!

This specific snippet we're discussing here isn't updating for some reason, but there is now logic to cover the case where a non-required parameter has a truthy value.  So yes, in the latest revision (at least as of rev https://github.com/natcap/invest/pull/1511/commits/617e76d9bbaaa2ecc66cf041b967d7fd2748a945) I would expect the validation you describe to take place.

Reply:
Ah okay, thanks for clarifying. I think that validation would _not_ take place before, because `excluded_keys` were removed from the set to validate: https://github.com/natcap/invest/blob/9a706487a1f8a8a9334435842f3c75f50fa7c590/src/natcap/invest/validation.py#L1003-L1004
Is that a change worth noting in history?

Reply:
Oh!  I see what you mean.  Sorry for my confusion about this!  Yes, this is indeed a change in behavior that I _think_ is for the best, but I will share a dev build with Stacie so that she can try it out.  In the meantime, I have noted the changes in `HISTORY`.